### PR TITLE
Add a no-modal-dialogs option.

### LINKED ISSUE #2006: Add a --no-modal-dialogs option (subject to agreed-on wording)
- At first, disable the first-run dialog box even if no prefs are specified
             , disable sudo

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #2006

### CODE REVIEW

**Comment by jandubois** [comment]
I don't quite understand the purpose of this option: when any commandline option is specified, RD assumes that any non-specified settings will have their default values. So there should be no first-run dialog.

I can see that for now it could enable `sudo` mode, so will prompt for a password unless the settings specify "no-root" or whatever we are going to call it. But once we install RD with a privileged helper process, the need for a sudo dialog should be gone, so at that point this option can also be removed again.

Please let me know if I misunderstand the situation!

**Comment by ericpromislow** [comment]
It's main use is if you don't care about any of the initial settings, but don't want any error messages to appear in a dialog box

**Comment by jandubois** [comment]
> don't want any error messages

Ah, yes, I forgot about that. My initial thought was that an API call should never cause a dialog to pop up, but I can see that if you use it for e.g. integration with an IDE, then this might still be ok.

So this should be a "transient" option on the "PUT settings" call, so both `rdctl start` and `rdctl` can pass it along, but it will not be stored with the engine, but only effective for this call.

**Comment by mook-as** [comment]
Would this make more sense in a future where we split the front end from the back end, so that we can run RD headless instead?

**Comment by jandubois** [comment]
> Would this make more sense in a future where we split the front end from the back end, so that we can run RD headless instead?

Aren't we almost there now?

```console
rdctl start --no-modal-dialogs --sudo=false --kubernetes-version 1.23
rdctl wait ...
# use your cluster
rdctl shutdown
```

So we probably don't have a `--sudo=false` option yet (and `rdctl wait` isn't implemented either), but otherwise this should work.

Now, we can argue if `--no-modal-dialogs` should imply `--sudo=false`, and we can also argue that we may want a `--hide` option to not display the windows by default.

**Comment by ericpromislow** [comment]
The split-fix commit ended up in this branch as well. It doesn't want to be deleted, so please hold off on reviewing this PR until PR #2040 gets approved and merged in.

File: src/main/commandServer/__tests__/settingsValidator.spec.ts
Comment (mook-as) [inline src/main/commandServer/__tests__/settingsValidator.spec.ts:169]: FYI, this can be destructured as well, if we do:
```ts
({ needToUpdate, needToRestart, errors } = subject.validateSettings(…));

**Review by mook-as — CHANGES_REQUESTED** [review]
So if the user uses `rdctl` once, then the app magically loses all its modal dialogs until the app next starts? That seems unexpected, especially if the user continues to use the UI in addition to the CLI.

File: src/config/settings.ts
Comment (jandubois) [reply src/config/settings.ts:236]: That's what I meant by "transient option" in https://github.com/rancher-sandbox/rancher-desktop/pull/2039#issuecomment-1099785619:

I expect (after the engine refactoring) that each `set` API call will create some kind of context that will be passed along internally until the desired state has been reached. The transient options ("no sudo", "no modal dialog") would override whatever permanent settings we have, but for this transaction only.

I understand that this may not be possible until after the engine refactoring.

File: src/config/settings.ts
Comment (ericpromislow) [reply src/config/settings.ts:163]: Yes, it does. I thought of just using `indexOf` and `substring` while working on this and will do that in the next commit.

File: src/config/settings.ts
Comment (ericpromislow) [reply src/config/settings.ts:236]: Yup, I wanted to avoid changing the interface of LimaBackend / K8s.KubernetesBackend but this would require such a transient setting.

File: src/config/settings.ts
Comment (ericpromislow) [reply src/config/settings.ts:253]: No, because I only needed to check it the first time we load the settings. In fact, `tray.ts` calls `settings.load` later, and if we put the check in `load` and had set it explicitly on the command-line, the second call would turn it off again.

File: src/config/__tests__/settings.spec.ts
Comment (ericpromislow) [reply src/config/__tests__/settings.spec.ts:93]: Done with a test:
```
expect(newArgs).not.toBe(originalArgs);
```

File: src/go/rdctl/cmd/set.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/set.go:63]: Because I read the documentation and saw that if they specify `--kubernetes-enabled` we want it to be `true`. Or did I get something else wrong?

File: src/main/commandServer/settingsValidator.ts
Comment (ericpromislow) [reply src/main/commandServer/settingsValidator.ts:144]: It's an alias only because `checkEnabled` doesn't need to modify the result of `checkBoolean`, while `checkNoModalDialogs` does. I suppose `checkEnabled` can get pulled out, and other fields that do the same check will just map to `checkBoolean`.

File: src/main/commandServer/__tests__/settingsValidator.spec.ts
Comment (ericpromislow) [reply src/main/commandServer/__tests__/settingsValidator.spec.ts:169]: So that's how it's done...

File: src/main/commandServer/settingsValidator.ts
Comment (mook-as) [reply src/main/commandServer/settingsValidator.ts:144]: Yeah, with the splat syntax we can make `checkNoModalDialogs` simple and everything else just use `checkBoolean` directly, I think.

File: src/go/rdctl/cmd/set.go
Comment (mook-as) [reply src/go/rdctl/cmd/set.go:63]: This is the default value (as in, if the flag was never explicitly specified by the user), right? Wouldn't that default to false? Otherwise we'd always want `--kubernetes-enabled=false` (as opposed to just not specifying it) if we don't want Kubernetes?

File: src/main/commandServer/settingsValidator.ts
Comment (ericpromislow) [reply src/main/commandServer/settingsValidator.ts]: I've reverted the changes to the settings validator in this PR because while they're good changes, they aren't needed for the PR's main function.

File: src/main/commandServer/settingsValidator.ts
Comment (ericpromislow) [reply src/main/commandServer/settingsValidator.ts]: This has all been reverted.

File: src/main/commandServer/__tests__/settingsValidator.spec.ts
Comment (ericpromislow) [reply src/main/commandServer/__tests__/settingsValidator.spec.ts]: But it's reverted for now.

File: src/main/commandServer/__tests__/settingsValidator.spec.ts
Comment (ericpromislow) [reply src/main/commandServer/__tests__/settingsValidator.spec.ts]: Also reverted for this PR.

File: src/go/rdctl/cmd/set.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/set.go:61]: Mark is correct -- it's the default value for the flag. But for the `set` and (soon to be PRed `start` command), we only process flags that are actually specified by consulting the `Changed` property. From the docs at https://pkg.go.dev/github.com/spf13/pflag#FlagSet.Changed:

```
Changed returns true if the flag was explicitly set during Parse() and false otherwise 
```

So better to be consistent, although those default values don't matter (so far).

File: src/config/settings.ts
Comment (ericpromislow) [reply src/config/settings.ts]: And the new PR implements a TransientSettings type of exactly one property.

File: src/main/commandServer/settingsValidator.ts
Comment (ericpromislow) [reply src/main/commandServer/settingsValidator.ts]: True, and it won't be in the eventual new 'n improved settingsValidator PR :)

**Comment by ericpromislow** [comment]
If it's not clear from above, the `noModalDialogs` setting is now transient. If the next run doesn't specify it, and there's a need for modal dialogs to appear, they will.

**Review by mook-as — CHANGES_REQUESTED** [review]
Talked with Eric to confirm what this is trying to do:
- Starting the main Rancher Desktop now takes a `--no-modal-dialogs` option, which suppresses modal dialogs:
  - sudo prompt: assume no sudo
  - error dialog: log & exit
  - first run: assume defaults
- This lasts for that run of Rancher Desktop (and is unrelated to the lifetime of `rdctl`)

Eric says he's spotted some issues and I should review after he pushes more changes.

**Comment by jandubois** [comment]
> * This lasts for that run of Rancher Desktop (and is unrelated to the lifetime of `rdctl`)

I would think that as soon as the user does a settings change (e.g. switch from `containerd` to `moby`, or ask for a Factory Reset), that this option would be dismissed, and they would still get a dialog asking for confirmation. Correct?

**Comment by ericpromislow** [comment]
Right now the `--no-modal-dialogs` setting doesn't affect those UI-side confirmations. It only affects dialog boxes that are not immediately triggered by a user action.

I think we're ok taking the simplest action, which is to keep things the way they are -- the setting prevents only dialogs that are not directly related to user actions from appearing and stopping workflow, and no user action can change the setting. If you want to change the setting, restart Rancher Desktop without it.

**Comment by jandubois** [comment]
> the setting prevents only dialogs that are not directly related to user actions from appearing and stopping workflow

I don't think this is quite true: when you switch e.g. the container runtime, you could get an error that would put up a dialog, and now it will just exit.

I still think this is fine for now, and we can revisit after the engine refactoring.

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/config/settings.ts
Comment (ericpromislow) [reply src/config/settings.ts]: I just renamed it -- it's not the default in the same sense as it is for the regular settings, where those values are typically used only the first time the app is used (or after a factory reset).

**Review by mook-as — DISMISSED** [review]

File: src/config/__tests__/settings.spec.ts
Comment (mook-as) [inline src/config/__tests__/settings.spec.ts:179]: (lint issue)

Comment (ericpromislow) [reply src/config/__tests__/settings.spec.ts]: done

**Review by mook-as — DISMISSED** [review]

**Review by jandubois — APPROVED** [review]
I dismissed @mook-as's approval when I had to rebase to resolve the merge conflict; adding approval again, but I did not independently review the code.

**Pull request merged**
